be/src/runtime/routine_load/routine_load_task_executor.cpp

```diff
@@ -307,6 +407,39 @@ void RoutineLoadTaskExecutor::exec_task(StreamLoadContext* ctx, DataConsumerPool
         };
         DeferOp delete_tp([tp_deleter] { return tp_deleter(); });
     } break;
+    case TLoadSourceType::PULSAR: {
+        for (auto& kv : ctx->pulsar_info->ack_offset) {
+            Status st;
+            // get consumer
+            std::shared_ptr<DataConsumer> consumer;
+            st = _data_consumer_pool.get_consumer(ctx, &consumer);
+            if (!st.ok()) {
+                // Pulsar Offset Acknowledgement is idempotent, Failure should not block the normal process
+                // So just print a warning
+                LOG(WARNING) << st.get_error_msg();
+                break;
+            }
+
+            // assign partition for consumer
+            st = std::static_pointer_cast<PulsarDataConsumer>(consumer)->assign_partition(kv.first, ctx);
+            if (!st.ok()) {
+                // Pulsar Offset Acknowledgement is idempotent, Failure should not block the normal process
```

Comment:
The messages may be consumed repeatedly if the BE break down.

Reply:
It's true, let's do exactly once in next PR.